nh_eobs_mental_health/models/nh_clinical_wardboard.py

```diff
@@ -259,75 +259,130 @@ def fields_view_get(self, cr, uid, view_id=None, view_type='form',
-                    ],
-                    context=context)
-                rec['rapid_tranq'] = spell.get('rapid_tranq')
-
-                if rec.get('next_blood_glucose_diff') \
-                        and rec.get('next_blood_glucose_diff') == '00:00':
-                    rec['next_blood_glucose_diff'] = ''
-                elif not rec.get('next_blood_glucose_diff'):
-                    rec['next_blood_glucose_diff'] = ''
-
-                if spell.get('obs_stop'):
-                    obs_stop_model = self.pool['nh.clinical.pme.obs_stop']
-                    obs_stops = obs_stop_model.search(cr, user, [
-                        ['spell', '=', spell_id]
-                    ], context=context)
-                    if obs_stops:
-                        obs_stop = obs_stops[0]
-                        reason = obs_stop_model.read(
-                            cr, user, obs_stop, ['reason'], context=context)
-                        rec['frequency'] = reason.get('reason', [0, False])[1]
-                    rec['next_diff'] = 'Observations Stopped'
-                    rec['next_blood_glucose_diff'] = 'Observations Stopped'
-                elif spell.get('refusing_obs'):
-                    rec['frequency'] = 'Refused - {0}'.format(rec['frequency'])
-                    rec['next_diff'] = 'Refused - {0}'.format(rec['next_diff'])
-                if not spell.get('obs_stop') and spell.get('refusing_obs_blood_glucose'):
-                    if rec.get('blood_glucose_frequency'):
-                        rec['blood_glucose_frequency'] = 'Refused - {0}'.format(rec['blood_glucose_frequency'])
-                        rec['next_blood_glucose_diff'] = 'Refused - {0}'.format(rec['next_blood_glucose_diff'])
-        if was_single_record:
-            return res[0]
+    def read(self, cr, user, ids, fields=None, context=None, load='_classic_read'):
+        res = super(NHClinicalWardboard, self).read(cr, user, ids, fields=fields, context=context, load=load)
+
+        patient_ids = [self._get_patient_id(record) for record in res]
+        obj_nh_clinical_spell = self.pool['nh.clinical.spell']
+        spell_ids = obj_nh_clinical_spell.search(cr, user, [
+                ('patient_id', 'in', patient_ids),
+                ('state', 'not in', ['completed', 'cancelled']),
+            ],
+            context=context
+        )
+        spells = obj_nh_clinical_spell.browse(cr, user, spell_ids, context=context)
+        for record in res:
+            spell = spells.filtered(
+                lambda _spell: _spell.patient_id.id == self._get_patient_id(record)
+            )
+            if spell:
+                if record.get('next_blood_glucose_diff', '00:00') == '00:00':
+                    record['next_blood_glucose_diff'] = ''
+                record['rapid_tranq'] = spell.rapid_tranq
+
+                if spell.obs_stop:
+                    record['frequency'] = self._get_stopped_obs_reason(
```

Comment:
This like is potentially going to be generating a lot of queries.

It looks more difficult to split out because it depends on certain conditions from spells, but there should be a way to batch this up.

Not saying it needs to be done in this PR but maybe can go back to it at some point if slowness is still an issue

Reply:
I don't know figures of how many patients will have obs stopped at any particular time, if performance becomes an issue with a large percentage of stopped obs then will revisit this.